### ISSUE TITLE
`_prepare_4d_attention_mask_for_sdpa` is not for causal attention but claims...

### ISSUE THREAD

**Issue by minostauros** [huggingface/transformers#30095]
... SDPA causal mask generation may be wrong for the mask generation.

https://github.com/huggingface/transformers/blob/76fa17c1663a0efeca7208c20579833365584889/src/transformers/modeling_attn_mask_utils.py#L421-L433

Will it be safe to just return `None` for the `else:` case?

For causal attention, we can just use `_prepare_4d_causal_attention_mask_for_sdpa`

Related issues:
https://github.com/pytorch/pytorch/issues/108108
https://github.com/Dao-AILab/flash-attention/commit/9e5e8bc91e30af5cdc321362b553f6c0da332e30
https://github.com/huggingface/transformers/pull/28802

**Comment by amyeroberts** [comment]
cc @fxmarty

**Comment by fxmarty** [comment]
Hi, here is a docstring for this function:

**Comment by minostauros** [comment]
> This is a somewhat unlikely case though, where one would use past key values for typically encoder-type of models. How did you run into this case?

I didn't run into the specific section but I was just reviewing #28802 and was trying to add flash-attention-2 to BERT (BLIP-2 variant of BERT to be exact).
Thanks for confirmation!

**Comment by minostauros** [comment]
This issue will be closed by https://github.com/huggingface/transformers/pull/30138
